### DISCUSSION TITLE
Cannot open serial port with PCI.

### DISCUSSION THREAD

**Discussion by noatshuva** [serialport/serialport-rs#145] [Q&A]
Hi,

I have a Quectel rm520n-gl modem with both USB and PCI.
I'm trying to open a serial port with both. USB works fine but with PCI I keep getting an "Invalid Arguments" error.
USB port name is `ttyUSB3` and PCI port name is `mhi_DUN`.
Running `serialport::available_ports()` outputs only USB and doesn't recognize PCI. 
Why can't I open the serial port with PCI?

thanks

**Comment by sirhcel** [comment]
Hello @noatshuva, 

> I have a Quectel rm520n-gl modem with both USB and PCI. I'm trying to open a serial port with both. USB works fine [...]

does this include that you are able to use the modem via the USB port (`/dev/ttyUSB3`)?

> [...] but with PCI I keep getting an "Invalid Arguments" error. USB port name is `ttyUSB3` and PCI port name is `mhi_DUN`. Running `serialport::available_ports()` outputs only USB and doesn't recognize PCI.

Where does this error come from? From `SerialPortBuilder::open`? What is the exact error type you are getting? Could you share your code up to the point where it fails with us? How are permissions on your system set up? What do the following commands report?

```
$ ls -l /dev/ttyUSB3 /dev/mhi_DUN
$ id
```

**Comment by noatshuva** [comment]
Hey @sirhcel,

> does this include that you are able to use the modem via the USB port (`/dev/ttyUSB3`)?

Yes. I am able to use the modem via USB port.

> Where does this error come from? From `SerialPortBuilder::open`? What is the exact error type you are getting? Could you share your code up to the point where it fails with us? How are permissions on your system set up? What do the following commands report?

Yes it comes from `SerialPortBuilder::open`. 
This is my code:

```
fn main() {
        println!("{:?}", serialport::available_ports());

        let mut usb_port = serialport::new("/dev/ttyUSB3", 115_200)
            .timeout(std::time::Duration::from_secs(2))
            .open()
            .expect("Failed to open port");
        usb_port.write_data_terminal_ready(true).ok();
        println!("Successfully connected to ttyUSB3");

        let mut pci_port = serialport::new("/dev/mhi_DUN", 115_200)
            .timeout(std::time::Duration::from_secs(2))
            .open()
            .expect("Failed to open port");
        pci_port.write_data_terminal_ready(true).ok();
        println!("Successfully connected to mhi_DUN");
}
```
> ```

Output:
```

Ok([SerialPortInfo { port_name: "/dev/ttyTHS0", port_type: Unknown }, SerialPortInfo { port_name: "/dev/ttyTHS1", port_type: Unknown }, SerialPortInfo { port_name: "/dev/ttyTHS3", port_type: Unknown }, SerialPortInfo { port_name: "/dev/ttyTHS4", port_type: Unknown }, SerialPortInfo { port_name: "/dev/ttyAMA0", port_type: Unknown }, SerialPortInfo { port_name: "/dev/ttyUSB0", port_type: UsbPort(UsbPortInfo { vid: 11388, pid: 2049, serial_number: Some("c15557a3"), manufacturer: Some("Quectel"), product: Some("RM520N-GL") }) }, SerialPortInfo { port_name: "/dev/ttyUSB1", port_type: UsbPort(UsbPortInfo { vid: 11388, pid: 2049, serial_number: Some("c15557a3"), manufacturer: Some("Quectel"), product: Some("RM520N-GL") }) }, SerialPortInfo { port_name: "/dev/ttyUSB2", port_type: UsbPort(UsbPortInfo { vid: 11388, pid: 2049, serial_number: Some("c15557a3"), manufacturer: Some("Quectel"), product: Some("RM520N-GL") }) }, SerialPortInfo { port_name: "/dev/ttyUSB3", port_type: UsbPort(UsbPortInfo { vid: 11388, pid: 2049, serial_number: Some("c15557a3"), manufacturer: Some("Quectel"), product: Some("RM520N-GL") }) }, SerialPortInfo { port_name: "/dev/ttyTCU0", port_type: Unknown }])
Successfully connected to ttyUSB3
thread 'main' panicked at src/main.rs:17:10:
Failed to open port: Error { kind: Unknown, description: "Invalid argument" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

 ls -l /dev/ttyUSB3 /dev/mhi_DUN output (I run my code with sudo): 
```

crw------- 1 root root    503, 3 Dec 21 08:36 /dev/mhi_DUN
crw-rw---- 1 root dialout 188, 3 Dec 21 13:26 /dev/ttyUSB3
```

id output:
`
uid=1000(ubuntu) gid=1000(ubuntu) groups=1000(ubuntu),4(adm),27(sudo),29(audio),44(video),103(render),130(gdm),996(weston-launch)`

**Reply by sirhcel** [reply]
I'm wondering what the invalid argument actually is. Could you please re-run your example with backtraces enabled? By setting `RUST_BACKTRACE` beforehand like:
```
$ export RUST_BACKTRACE=1
```

**Comment by noatshuva** [comment]
Also, I've noticed that the timeout doesn't work for me when reading. 
I have the following function to send and read at-commands:

```
fn send_at_command(port: &mut dyn SerialPort, command: &str) -> Result<String, String> {
    println!("Write command: {command}");
    port.set_timeout(std::time::Duration::from_millis(500)).ok();
    // Send AT command
    port.write_all(command.as_bytes())
        .map_err(|e| format!("Failed to write {command}: {}", e.to_string()))?;

    port.flush().map_err(|e| format!("Failed to run flush"))?;

    let mut buffer = vec![0; 256];
    port.read(&mut buffer)
        .map_err(|e| format!("Failed to read {command}: {}", e.to_string()))?;
    println!("Response: {}", String::from_utf8_lossy(&buffer).to_string());
    Ok(String::from_utf8_lossy(&buffer).to_string())
}
```

For some reason sometimes this function gets stuck in 'port.read'. Any idea why?

**Reply by sirhcel** [reply]
Strange. The timeout you are setting should be active for your call to `read`. I have to admit that i have no clue in the first place. Which device and driver are you using here? Could you run your program with `strace`? The timeout should be implemented through through the `ppoll` syscall. Syscall tracing might bring up more insights here.
